### PR TITLE
Format foreign functions and statics

### PR DESCRIPTION
**Pull request by marcusklaas**
Fixes https://github.com/nrc/rustfmt/issues/299.

Blocked by https://github.com/rust-lang/rust/issues/28472.

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -87,15 +96,99 @@ impl<'a> FmtVisitor<'a> {
...
+    pub fn format_foreign_mod(&mut self, fm: &ast::ForeignMod, span: Span) {
+        self.buffer.push_str("extern ");
+
+        if fm.abi != abi::Abi::C {
+            self.buffer.push_str(&format!("{} ", fm.abi));
+        }
+
+        let snippet = self.snippet(span);
+        let brace_pos = snippet.find_uncommented("{").unwrap() as u32;
+        self.last_pos = span.lo + BytePos(brace_pos);
```
Comment (cassiersg) [inline src/items.rs]: This skips comments.

File: src/items.rs [hunk 2]
```diff
@@ -87,15 +96,99 @@ impl<'a> FmtVisitor<'a> {
...
+                                                   span,
+                                                   false);
+
+                match rewrite {
+                    Some(new_fn) => {
+                        self.buffer.push_str(format_visibility(item.vis));
+                        self.buffer.push_str(&new_fn);
+                        self.buffer.push_str(";");
+                    }
+                    None => return,
```
Comment (cassiersg) [inline src/items.rs]: A format_missing untill item.span.hi would be better, I think.

File: src/items.rs [hunk 3]
```diff
@@ -87,15 +96,99 @@ impl<'a> FmtVisitor<'a> {
...
+                    ""
+                };
+                let prefix = format!("{}static {}{}: ",
+                                     format_visibility(item.vis),
+                                     mut_str,
+                                     item.ident);
+                let offset = self.block_indent + prefix.len();
+                // 1 = ;
+                let width = self.config.max_width - offset - 1;
+                let rewrite = ty.rewrite(&self.get_context(), width, offset);
```
Comment (cassiersg) [inline src/items.rs]: Also possible comment loss between keywords.

File: src/items.rs [hunk 4]
```diff
@@ -87,15 +96,99 @@ impl<'a> FmtVisitor<'a> {
...
+
+                match rewrite {
+                    Some(result) => {
+                        self.buffer.push_str(&prefix);
+                        self.buffer.push_str(&result);
+                        self.buffer.push_str(";");
+                    }
+                    None => return,
```
Comment (cassiersg) [inline src/items.rs]: Idem, format_missing.

**Comment by cassiersg** [comment]
Looks good !
For the comment loss in crazy places, I think we should detect if there are comments in a given span, and then not reformat it there are (assuming the user knows what he does with comments there).
You may also just put a fixme (I want to experience with this way of doing, to remove all sources of comment loss in rustfmt.)

File: src/items.rs (hunk 2 above)
Comment (marcusklaas) [inline src/items.rs]: Isn't that functionally equivalent to the current situation?

**Comment by marcusklaas** [comment]
Comments addressed. I added a FIXME for the missing comments in the function declaration so that this can be dealt with in one fell swoop when its time comes.

Thanks for the review @cassiersg!

**Pull request merged**